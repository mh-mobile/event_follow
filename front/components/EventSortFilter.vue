<template lang="pug">
  .example-modal-window
    .event_header_sort_filter(@click="openModal")
      | {{ getCurrentConditionLabel }}

    Modal(@close="closeModal" v-if="modal")
      .event-modal-container
        form(action="/events" method="GET" id="js-sort-filter-form")
          .event-modal-header
            .event-modal-left
              | Sort by
            .event-modal-center
            .event-modal-right
              | Filter by
          .event-modal-content
              .event-modal-left.event-modal-sort
                select(v-model="selectedSortCondition" @change="selectedSortConditionChanged($event)" name="sort")
                  option(v-for="item in eventSortConditionItems" :value="item.value" :key="item.value")
                    | {{ item.name }}
              .event-modal-center
                | ×
              .event-modal-right.event-modal-sort
                select(v-model="selectedTimeFilterCondition" v-show="isFrinedsNumberSortCondition" @change="selectedTimeFilterConditionChanged" name="time")
                  option(v-for="item in time_filter_type_items" :value="item.value" :key="item.value")
                    | {{ item.name }}
                select(v-model="selectedFriendsFilterCondition" v-show="!isFrinedsNumberSortCondition" @change="selectedFriendsFilterConditionChanged" name="friends")
                  option(v-for="item in friends_filter_type_items" :value="item.value" :key="item.value")
                    | {{ item.name }}
  </div>
</template>

<script>
import Modal from "@/components/Modal.vue"
import {
  defineComponent,
  computed,
  reactive,
  toRefs,
  onMounted,
  watch
} from "@nuxtjs/composition-api"

export default defineComponent({
  components: { Modal },
  props: {
    eventSortType: {
      type: String,
      default: ""
    },
    timeFilterType: {
      type: String,
      default: ""
    },
    friendsFilterType: {
      type: String,
      default: ""
    }
  },
  setup(props, { root }) {
    const eventSortConditionItems = [
      {
        name: "Friend数",
        value: "friends_number_order"
      },
      {
        name: "新着順",
        value: "recent_order"
      },
      {
        name: "投稿順",
        value: "created_order"
      },
      {
        name: "開催が近い順",
        value: "closeness_order"
      }
    ]

    const time_filter_type_items = [
      {
        name: "過去8時間",
        value: "past_8_hours"
      },
      {
        name: "過去24時間",
        value: "past_24_hours"
      },
      {
        name: "過去2日",
        value: "past_2_days"
      },
      {
        name: "過去3日",
        value: "past_3_days"
      },
      {
        name: "過去4日",
        value: "past_4_days"
      },
      {
        name: "過去5日",
        value: "past_5_days"
      },
      {
        name: "過去6日",
        value: "past_6_days"
      },
      {
        name: "過去1週間",
        value: "past_1_weeks"
      },
      {
        name: "All",
        value: "past_all"
      }
    ]

    const friends_filter_type_items = [
      {
        name: "Friends 1+",
        value: "one_or_more_friends"
      },
      {
        name: "Friends 2+",
        value: "two_or_more_friends"
      },
      {
        name: "Friends 3+",
        value: "three_or_more_friends"
      },
      {
        name: "Friends 4+",
        value: "four_or_more_friends"
      },
      {
        name: "Friends 5+",
        value: "five_or_more_friends"
      }
    ]

    const state = reactive({
      modal: false,
      message: "",
      selectedSortCondition: "friends_number_order",
      selectedTimeFilterCondition: "past_8_hours",
      selectedFriendsFilterCondition: "one_or_more_friends"
    })

    const isFrinedsNumberSortCondition = computed(() => {
      return state.selectedSortCondition === "friends_number_order"
    })
    const getCurrentConditionLabel = computed(() => {
      if (
        state.selectedSortCondition === "" ||
        state.selectedTimeFilterCondition === "" ||
        state.selectedFriendsFilterCondition === ""
      ) {
        return ""
      }

      const sortConditionLabel = eventSortConditionItems.find((item) => {
        return item.value === state.selectedSortCondition
      }).name
      if (state.selectedSortCondition === "friends_number_order") {
        const timeFilterConditionLabel = time_filter_type_items.find((item) => {
          return item.value === state.selectedTimeFilterCondition
        }).name
        return `${sortConditionLabel} × ${timeFilterConditionLabel}`
      } else {
        const friendsFilterConditionLabel = friends_filter_type_items.find(
          (item) => {
            return item.value === state.selectedFriendsFilterCondition
          }
        ).name
        return `${sortConditionLabel} × ${friendsFilterConditionLabel}`
      }
    })

    const openModal = () => {
      state.modal = true
    }

    const closeModal = () => {
      state.modal = false
    }

    const selectedSortConditionChanged = () => {
      root.$router.replace(
        `/events?sort=${state.selectedSortCondition}&time=${state.selectedTimeFilterCondition}&friends=${state.selectedFriendsFilterCondition}`
      )
      state.modal = false
    }

    const selectedTimeFilterConditionChanged = () => {
      root.$router.replace(
        `/events?sort=${state.selectedSortCondition}&time=${state.selectedTimeFilterCondition}&friends=${state.selectedFriendsFilterCondition}`
      )
      state.modal = false
    }

    const selectedFriendsFilterConditionChanged = () => {
      root.$router.replace(
        `/events?sort=${state.selectedSortCondition}&time=${state.selectedTimeFilterCondition}&friends=${state.selectedFriendsFilterCondition}`
      )
      state.modal = false
    }

    watch(
      () => props.eventSortType,
      async (newValue, oldValue) => {
        state.selectedSortCondition = newValue
      }
    )

    watch(
      () => props.timeFilterType,
      async (newValue, oldValue) => {
        state.selectedTimeFilterCondition = newValue
      }
    )

    watch(
      () => props.friendsFilterType,
      async (newValue, oldValue) => {
        state.selectedFriendsFilterCondition = newValue
      }
    )

    return {
      ...toRefs(state),
      eventSortConditionItems,
      time_filter_type_items,
      friends_filter_type_items,
      isFrinedsNumberSortCondition,
      getCurrentConditionLabel,
      openModal,
      closeModal,
      selectedSortConditionChanged,
      selectedTimeFilterConditionChanged,
      selectedFriendsFilterConditionChanged
    }
  }
})
</script>

<style lang="scss" scoped>
.event_header_sort_filter {
  font-size: 1.5em;
  font-weight: bold;
  position: relative;
  padding-right: 24px;
  cursor: pointer;

  &::after {
    content: "";
    position: absolute;
    width: 12px;
    height: 12px;
    border: 2px solid;
    top: calc(50% - 4px);
    right: 0;
    border-color: transparent transparent #565656 #565656;
    transform: rotate(-45deg) translateY(-50%);
  }
}

.event-modal-container {
  max-width: 30em;
  min-width: 30em;
  .event-modal-header {
    display: flex;
    width: 100%;
    align-items: center;
    justify-content: center;
    font-size: 1.2em;
    font-weight: bold;
    color: #9f9fa3;
  }
  .event-modal-content {
    display: flex;
    width: 100%;
    align-items: center;
    justify-content: center;
    height: 70px;
  }
  .event-modal-left {
    width: 45%;
    display: flex;
    justify-content: center;
  }
  .event-modal-center {
    width: 10%;
    display: flex;
    justify-content: center;
    font-size: 2em;
  }
  .event-modal-right {
    width: 45%;
    display: flex;
    justify-content: center;
  }
  .event-modal-sort,
  .event-modal-filter {
    position: relative;
    &::after {
      content: "";
      position: absolute;
      width: 12px;
      height: 12px;
      border: 2px solid;
      top: calc(50% - 4px);
      right: 5px;
      border-color: transparent transparent #565656 #565656;
      transform: rotate(-45deg) translateY(-50%);
    }
    select {
      width: 100%;
      height: 50px;
      padding-left: 10px;
      padding-right: 10px;
      outline: none;
      -webkit-appearance: none;
      -moz-appearance: none;
      appearance: none;
      font-size: 1em;
      border-radius: 5px;
    }
  }
}
</style>
