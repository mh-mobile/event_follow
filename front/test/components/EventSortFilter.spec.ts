import { createLocalVue, mount } from "@vue/test-utils"
import VueCompositionApi from "@vue/composition-api"
import EventSortFilter from "@/components/EventSortFilter.vue"

const localVue = createLocalVue()
localVue.use(VueCompositionApi)
describe("EventSortFilter.vue", () => {
  it("sort=friends_number_order&time=past_8_hours", async () => {
    const wrapper = mount(EventSortFilter, {
      localVue,
      propsData: {
        eventSortType: "friends_number_order",
        friendsFilterType: "three_or_more_friends"
      }
    })
    await wrapper.find(".event_header_sort_filter").trigger("click")
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("sort=friends_number_order&time=past_24_hours", async () => {
    const wrapper = mount(EventSortFilter, {
      localVue,
      propsData: {
        eventSortType: "friends_number_order",
        friendsFilterType: "past_24_hours"
      }
    })
    await wrapper.find(".event_header_sort_filter").trigger("click")
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("sort=recent_order&friends=one_or_more_friends", async () => {
    const wrapper = mount(EventSortFilter, {
      localVue,
      propsData: {
        eventSortType: "recent_order",
        friendsFilterType: "one_or_more_friends"
      }
    })
    await wrapper.find(".event_header_sort_filter").trigger("click")
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("sort=created_order&friends=two_or_more_friends", async () => {
    const wrapper = mount(EventSortFilter, {
      localVue,
      propsData: {
        eventSortType: "created_order",
        friendsFilterType: "two_or_more_friends"
      }
    })
    await wrapper.find(".event_header_sort_filter").trigger("click")
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("sort=closeness_order&friends=two_or_more_friends", async () => {
    const wrapper = mount(EventSortFilter, {
      localVue,
      propsData: {
        eventSortType: "closeness_order",
        friendsFilterType: "two_or_more_friends"
      }
    })
    await wrapper.find(".event_header_sort_filter").trigger("click")
    expect(wrapper.html()).toMatchSnapshot()
  })
})
