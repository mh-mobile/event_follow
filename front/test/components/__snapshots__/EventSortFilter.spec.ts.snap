// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventSortFilter.vue sort=closeness_order&friends=two_or_more_friends 1`] = `
"<div class=\\"example-modal-window\\">
  <div class=\\"event_header_sort_filter\\">開催が近い順 × Friends 2+</div>
  <transition-stub name=\\"modal\\" appear=\\"true\\">
    <div class=\\"modal modal-overlay\\">
      <div class=\\"modal-window\\">
        <div class=\\"modal-window-content\\">
          <div class=\\"event-modal-container\\">
            <form action=\\"/events\\" method=\\"GET\\" id=\\"js-sort-filter-form\\">
              <div class=\\"event-modal-sort\\">
                <div class=\\"event-modal-label\\">Sort by</div>
                <div class=\\"event-modal-value\\"><select name=\\"sort\\">
                    <option value=\\"friends_number_order\\">Friend数</option>
                    <option value=\\"recent_order\\">新着順</option>
                    <option value=\\"created_order\\">投稿順</option>
                    <option value=\\"closeness_order\\">開催が近い順</option>
                  </select></div>
              </div>
              <div class=\\"event-modal-center\\">
                <div class=\\"event-modal-label\\"></div>
                <div class=\\"event-modal-value\\">×</div>
              </div>
              <div class=\\"event-modal-filter\\">
                <div class=\\"event-modal-label\\">Filter by</div>
                <div class=\\"event-modal-value\\"><select name=\\"time\\" style=\\"display: none;\\">
                    <option value=\\"past_8_hours\\">過去8時間</option>
                    <option value=\\"past_24_hours\\">過去24時間</option>
                    <option value=\\"past_2_days\\">過去2日</option>
                    <option value=\\"past_3_days\\">過去3日</option>
                    <option value=\\"past_4_days\\">過去4日</option>
                    <option value=\\"past_5_days\\">過去5日</option>
                    <option value=\\"past_6_days\\">過去6日</option>
                    <option value=\\"past_1_weeks\\">過去1週間</option>
                    <option value=\\"past_all\\">All</option>
                  </select><select name=\\"friends\\">
                    <option value=\\"one_or_more_friends\\">Friends 1+</option>
                    <option value=\\"two_or_more_friends\\">Friends 2+</option>
                    <option value=\\"three_or_more_friends\\">Friends 3+</option>
                    <option value=\\"four_or_more_friends\\">Friends 4+</option>
                    <option value=\\"five_or_more_friends\\">Friends 5+</option>
                  </select></div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>"
`;

exports[`EventSortFilter.vue sort=created_order&friends=two_or_more_friends 1`] = `
"<div class=\\"example-modal-window\\">
  <div class=\\"event_header_sort_filter\\">投稿順 × Friends 2+</div>
  <transition-stub name=\\"modal\\" appear=\\"true\\">
    <div class=\\"modal modal-overlay\\">
      <div class=\\"modal-window\\">
        <div class=\\"modal-window-content\\">
          <div class=\\"event-modal-container\\">
            <form action=\\"/events\\" method=\\"GET\\" id=\\"js-sort-filter-form\\">
              <div class=\\"event-modal-sort\\">
                <div class=\\"event-modal-label\\">Sort by</div>
                <div class=\\"event-modal-value\\"><select name=\\"sort\\">
                    <option value=\\"friends_number_order\\">Friend数</option>
                    <option value=\\"recent_order\\">新着順</option>
                    <option value=\\"created_order\\">投稿順</option>
                    <option value=\\"closeness_order\\">開催が近い順</option>
                  </select></div>
              </div>
              <div class=\\"event-modal-center\\">
                <div class=\\"event-modal-label\\"></div>
                <div class=\\"event-modal-value\\">×</div>
              </div>
              <div class=\\"event-modal-filter\\">
                <div class=\\"event-modal-label\\">Filter by</div>
                <div class=\\"event-modal-value\\"><select name=\\"time\\" style=\\"display: none;\\">
                    <option value=\\"past_8_hours\\">過去8時間</option>
                    <option value=\\"past_24_hours\\">過去24時間</option>
                    <option value=\\"past_2_days\\">過去2日</option>
                    <option value=\\"past_3_days\\">過去3日</option>
                    <option value=\\"past_4_days\\">過去4日</option>
                    <option value=\\"past_5_days\\">過去5日</option>
                    <option value=\\"past_6_days\\">過去6日</option>
                    <option value=\\"past_1_weeks\\">過去1週間</option>
                    <option value=\\"past_all\\">All</option>
                  </select><select name=\\"friends\\">
                    <option value=\\"one_or_more_friends\\">Friends 1+</option>
                    <option value=\\"two_or_more_friends\\">Friends 2+</option>
                    <option value=\\"three_or_more_friends\\">Friends 3+</option>
                    <option value=\\"four_or_more_friends\\">Friends 4+</option>
                    <option value=\\"five_or_more_friends\\">Friends 5+</option>
                  </select></div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>"
`;

exports[`EventSortFilter.vue sort=friends_number_order&time=past_8_hours 1`] = `
"<div class=\\"example-modal-window\\">
  <div class=\\"event_header_sort_filter\\"></div>
  <transition-stub name=\\"modal\\" appear=\\"true\\">
    <div class=\\"modal modal-overlay\\">
      <div class=\\"modal-window\\">
        <div class=\\"modal-window-content\\">
          <div class=\\"event-modal-container\\">
            <form action=\\"/events\\" method=\\"GET\\" id=\\"js-sort-filter-form\\">
              <div class=\\"event-modal-sort\\">
                <div class=\\"event-modal-label\\">Sort by</div>
                <div class=\\"event-modal-value\\"><select name=\\"sort\\">
                    <option value=\\"friends_number_order\\">Friend数</option>
                    <option value=\\"recent_order\\">新着順</option>
                    <option value=\\"created_order\\">投稿順</option>
                    <option value=\\"closeness_order\\">開催が近い順</option>
                  </select></div>
              </div>
              <div class=\\"event-modal-center\\">
                <div class=\\"event-modal-label\\"></div>
                <div class=\\"event-modal-value\\">×</div>
              </div>
              <div class=\\"event-modal-filter\\">
                <div class=\\"event-modal-label\\">Filter by</div>
                <div class=\\"event-modal-value\\"><select name=\\"time\\">
                    <option value=\\"past_8_hours\\">過去8時間</option>
                    <option value=\\"past_24_hours\\">過去24時間</option>
                    <option value=\\"past_2_days\\">過去2日</option>
                    <option value=\\"past_3_days\\">過去3日</option>
                    <option value=\\"past_4_days\\">過去4日</option>
                    <option value=\\"past_5_days\\">過去5日</option>
                    <option value=\\"past_6_days\\">過去6日</option>
                    <option value=\\"past_1_weeks\\">過去1週間</option>
                    <option value=\\"past_all\\">All</option>
                  </select><select name=\\"friends\\" style=\\"display: none;\\">
                    <option value=\\"one_or_more_friends\\">Friends 1+</option>
                    <option value=\\"two_or_more_friends\\">Friends 2+</option>
                    <option value=\\"three_or_more_friends\\">Friends 3+</option>
                    <option value=\\"four_or_more_friends\\">Friends 4+</option>
                    <option value=\\"five_or_more_friends\\">Friends 5+</option>
                  </select></div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>"
`;

exports[`EventSortFilter.vue sort=friends_number_order&time=past_24_hours 1`] = `
"<div class=\\"example-modal-window\\">
  <div class=\\"event_header_sort_filter\\"></div>
  <transition-stub name=\\"modal\\" appear=\\"true\\">
    <div class=\\"modal modal-overlay\\">
      <div class=\\"modal-window\\">
        <div class=\\"modal-window-content\\">
          <div class=\\"event-modal-container\\">
            <form action=\\"/events\\" method=\\"GET\\" id=\\"js-sort-filter-form\\">
              <div class=\\"event-modal-sort\\">
                <div class=\\"event-modal-label\\">Sort by</div>
                <div class=\\"event-modal-value\\"><select name=\\"sort\\">
                    <option value=\\"friends_number_order\\">Friend数</option>
                    <option value=\\"recent_order\\">新着順</option>
                    <option value=\\"created_order\\">投稿順</option>
                    <option value=\\"closeness_order\\">開催が近い順</option>
                  </select></div>
              </div>
              <div class=\\"event-modal-center\\">
                <div class=\\"event-modal-label\\"></div>
                <div class=\\"event-modal-value\\">×</div>
              </div>
              <div class=\\"event-modal-filter\\">
                <div class=\\"event-modal-label\\">Filter by</div>
                <div class=\\"event-modal-value\\"><select name=\\"time\\">
                    <option value=\\"past_8_hours\\">過去8時間</option>
                    <option value=\\"past_24_hours\\">過去24時間</option>
                    <option value=\\"past_2_days\\">過去2日</option>
                    <option value=\\"past_3_days\\">過去3日</option>
                    <option value=\\"past_4_days\\">過去4日</option>
                    <option value=\\"past_5_days\\">過去5日</option>
                    <option value=\\"past_6_days\\">過去6日</option>
                    <option value=\\"past_1_weeks\\">過去1週間</option>
                    <option value=\\"past_all\\">All</option>
                  </select><select name=\\"friends\\" style=\\"display: none;\\">
                    <option value=\\"one_or_more_friends\\">Friends 1+</option>
                    <option value=\\"two_or_more_friends\\">Friends 2+</option>
                    <option value=\\"three_or_more_friends\\">Friends 3+</option>
                    <option value=\\"four_or_more_friends\\">Friends 4+</option>
                    <option value=\\"five_or_more_friends\\">Friends 5+</option>
                  </select></div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>"
`;

exports[`EventSortFilter.vue sort=recent_order&friends=one_or_more_friends 1`] = `
"<div class=\\"example-modal-window\\">
  <div class=\\"event_header_sort_filter\\">新着順 × Friends 1+</div>
  <transition-stub name=\\"modal\\" appear=\\"true\\">
    <div class=\\"modal modal-overlay\\">
      <div class=\\"modal-window\\">
        <div class=\\"modal-window-content\\">
          <div class=\\"event-modal-container\\">
            <form action=\\"/events\\" method=\\"GET\\" id=\\"js-sort-filter-form\\">
              <div class=\\"event-modal-sort\\">
                <div class=\\"event-modal-label\\">Sort by</div>
                <div class=\\"event-modal-value\\"><select name=\\"sort\\">
                    <option value=\\"friends_number_order\\">Friend数</option>
                    <option value=\\"recent_order\\">新着順</option>
                    <option value=\\"created_order\\">投稿順</option>
                    <option value=\\"closeness_order\\">開催が近い順</option>
                  </select></div>
              </div>
              <div class=\\"event-modal-center\\">
                <div class=\\"event-modal-label\\"></div>
                <div class=\\"event-modal-value\\">×</div>
              </div>
              <div class=\\"event-modal-filter\\">
                <div class=\\"event-modal-label\\">Filter by</div>
                <div class=\\"event-modal-value\\"><select name=\\"time\\" style=\\"display: none;\\">
                    <option value=\\"past_8_hours\\">過去8時間</option>
                    <option value=\\"past_24_hours\\">過去24時間</option>
                    <option value=\\"past_2_days\\">過去2日</option>
                    <option value=\\"past_3_days\\">過去3日</option>
                    <option value=\\"past_4_days\\">過去4日</option>
                    <option value=\\"past_5_days\\">過去5日</option>
                    <option value=\\"past_6_days\\">過去6日</option>
                    <option value=\\"past_1_weeks\\">過去1週間</option>
                    <option value=\\"past_all\\">All</option>
                  </select><select name=\\"friends\\">
                    <option value=\\"one_or_more_friends\\">Friends 1+</option>
                    <option value=\\"two_or_more_friends\\">Friends 2+</option>
                    <option value=\\"three_or_more_friends\\">Friends 3+</option>
                    <option value=\\"four_or_more_friends\\">Friends 4+</option>
                    <option value=\\"five_or_more_friends\\">Friends 5+</option>
                  </select></div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>"
`;
